validator/src/main.rs

```diff
@@ -0,0 +1,261 @@
+            .await
+            .expect("Failed to establish connection");
+        info!("Accepted connection");
+        let rng_resp = resp.clone();
+        let _ = tokio::spawn(
+            async move {
+                let resp_bytes = rng_resp.as_bytes();
+                // We expect this write to complete instantaneously,
+                // a timeout is not needed here.
+                match stream.write_all(resp_bytes).await {
+                    Ok(()) => debug!(written_bytes = resp_bytes.len()),
+                    Err(error) => error!(%error, "Failed to write bytes to client"),
+                }
+            }
+            .instrument(tracing::info_span!("conn", %client_addr)),
+        );
+    }
+}
+
+pub fn parse_timeout(s: &str) -> Result<time::Duration> {
+    let s = s.trim();
+    let (magnitude, unit) = if let Some(offset) = s.rfind(|c: char| c.is_ascii_digit()) {
+        let (magnitude, unit) = s.split_at(offset + 1);
+        let magnitude = magnitude.parse::<u64>()?;
+        (magnitude, unit)
+    } else {
+        anyhow::bail!("{} does not contain a timeout duration value", s);
+    };
+
+    let mul = match unit {
+        "" if magnitude == 0 => 0,
+        "ms" => 1,
+        "s" => 1000,
+        "m" => 1000 * 60,
+        "h" => 1000 * 60 * 60,
+        "d" => 1000 * 60 * 60 * 24,
+        _ => anyhow::bail!(
+            "invalid duration unit {} (expected one of 'ms', 's', 'm', 'h', or 'd')",
+            unit
+        ),
+    };
+
+    let ms = magnitude
+        .checked_mul(mul)
+        .ok_or_else(|| anyhow!("Timeout value {} overflows when converted to 'ms'", s))?;
+    Ok(time::Duration::from_millis(ms))
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::parse_timeout;
+    use std::time;
+
+    #[test]
+    fn test_parse_timeout_invalid() {
+        assert!(parse_timeout("120").is_err());
+        assert!(parse_timeout("s").is_err());
+        assert!(parse_timeout("foobars").is_err());
+        assert!(parse_timeout("18446744073709551615s").is_err())
+    }
+
+    #[test]
+    fn test_parse_timeout_seconds() {
+        assert_eq!(time::Duration::from_secs(0), parse_timeout("0").unwrap());
+        assert_eq!(time::Duration::from_secs(0), parse_timeout("0ms").unwrap());
+        assert_eq!(time::Duration::from_secs(0), parse_timeout("0s").unwrap());
+        assert_eq!(time::Duration::from_secs(0), parse_timeout("0m").unwrap());
+
+        assert_eq!(
+            time::Duration::from_secs(120),
+            parse_timeout("120s").unwrap()
+        );
+        assert_eq!(
+            time::Duration::from_secs(120),
+            parse_timeout("120000ms").unwrap()
+        );
+        assert_eq!(time::Duration::from_secs(120), parse_timeout("2m").unwrap());
+        assert_eq!(
+            time::Duration::from_secs(7200),
+            parse_timeout("2h").unwrap()
+        );
+        assert_eq!(
+            time::Duration::from_secs(172800),
+            parse_timeout("2d").unwrap()
+        );
```

Comment:
nit, take it or leave it: instead of `unwrap()`ping these, i might just change the assertions to 
```suggestion
        assert_eq!(Ok(time::Duration::from_secs(0)), parse_timeout("0"));
        assert_eq!(Ok(time::Duration::from_secs(0)), parse_timeout("0ms"));
        assert_eq!(Ok(time::Duration::from_secs(0)), parse_timeout("0s"));
        assert_eq!(Ok(time::Duration::from_secs(0)), parse_timeout("0m"));

        assert_eq!(
            Ok(time::Duration::from_secs(120)),
            parse_timeout("120s"),
        );
        assert_eq!(
            Ok(time::Duration::from_secs(120)),
            parse_timeout("120000ms"),
        );
        assert_eq!(Ok(time::Duration::from_secs(120)), parse_timeout("2m"));
        assert_eq!(
            Ok(time::Duration::from_secs(7200)),
            parse_timeout("2h"),
        );
        assert_eq!(
            Ok(time::Duration::from_secs(172800)),
            parse_timeout("2d"),
        );
```
that way, if any of these returns an error, the error message will be printed as part of the assertion output, so we also see the expected duration to compare against.

not a big deal though.

Reply:
I'd rather we print the error message as part of the assertion output, so this is a great suggestion :D